diff --git a/src/brpc/server.cpp b/src/brpc/server.cpp
index 9d938c6b..6c327f89 100644
--- a/src/brpc/server.cpp
+++ b/src/brpc/server.cpp
@@ -1889,7 +1889,7 @@ int StartDummyServerAt(int port, ProfilerLinker) {
                         "DummyServerOf(%s)", GetProgramName()));
             ServerOptions options;
             options.num_threads = 0;
-            if (dummy_server->Start(port, &options) != 0) {
+            if (dummy_server->Start(butil::string_printf("0.0.0.0:%d", port).c_str(), &options) != 0) {
                 LOG(ERROR) << "Fail to start dummy_server at port=" << port;
                 return -1;
             }
